### PR TITLE
Clarify that attribute keys are case-sensitive

### LINKED ISSUE #3782: Span attribute keys have unspecified case sensitivity
**What needs to be changed?** Describe the update that is required.

The documentation does not state whether span attribute keys are case sensitive

**What is the name + path of the page that needs changed?** The relative path
and page title where you found a problem.

concepts/signals/traces/#attributes - Span Attributes

**Additional context**: anything else you can think about that would be helpful.

If attribute keys are meant to be case sensitive - in line with most common programming languages - , any vendor that treats them case insensitive will cause a collision. The other way around an intentional case variation will result in a mismatch.

### PR DESCRIPTION
**Pull request by arminru**
Fixes #3782.

### CODE REVIEW

File: specification/common/README.md [hunk 1]
```diff
@@ -28,6 +28,7 @@ aliases: [/docs/reference/specification/common/common]
 An `Attribute` is a key-value pair, which MUST have the following properties:
 
 - The attribute key MUST be a non-`null` and non-empty string.
+  - Attribute keys are case-sensitive.
```
Comment (MrAlias) [inline specification/common/README.md:31]: Can we universally say this? I can seem many vendors or systems treating these as case-insensitive.

I think the most we can say is they need to be stored in an encoded form that preserves case-sensitivity.

File: specification/common/README.md [hunk 2]
```diff
@@ -28,6 +28,7 @@ aliases: [/docs/reference/specification/common/common]
...
 - The attribute value is either:
   - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
```
Comment (MrAlias) [inline specification/common/README.md:33]: A similar clarification will need to be made about string values as well.

**Review by MrAlias — COMMENTED** [review]
Have you verified with all the stable releases that this will not conflict with their implemenatations?

File: specification/common/README.md (hunk 1 above)
Comment (jack-berg) [reply specification/common/README.md:31]: I think yes. As @pyohannes points out [here](https://github.com/open-telemetry/opentelemetry-specification/issues/3782#issuecomment-1839216273):

> Attributes are defined as [Unicode sequences](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/). And different Unicode sequences would be different attribute names.

So this should just be a clarification. If a vendor treats them as case insensitive, they're doing so at their own risk.

File: specification/common/README.md (hunk 2 above)
Comment (reyang) [reply specification/common/README.md:33]: I have a general question - do we really want to add explicit clarification on all strings? I was under the impression that **people shouldn't assume strings are case insensitive unless it is called out explicitly**.

Comment (MrAlias) [reply specification/common/README.md:33]: Seems reasonable to me ^

File: specification/common/README.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/common/README.md:31]: I think we have always implicitly assumed and treated attribute keys as case sensitive. I am not aware of any Otel code that treats them differently. That the spec does not say so I would treat as a spec bug and this change is a bugfix.

**Review by tigrannajaryan — APPROVED** [review]

**Review by yurishkuro — APPROVED** [review]

File: specification/common/README.md (hunk 1 above)
Comment (MrAlias) [reply specification/common/README.md:31]: > I think we have always implicitly assumed and treated attribute keys as case sensitive. I am not aware of any Otel code that treats them differently. That the spec does not say so I would consider as a spec bug and this change is a bugfix.

I agree. But I don't think the added text is addressing what you are stating. We have always encode and treated attribute keys as case-sensitive[^*]. We have not always assumed attribute keys are extrinsically case-sensitive, there is nothing stopping other interpretations external to OTel.

[^*]: this hasn't been verified yet in all implementations

Comment (yurishkuro) [reply specification/common/README.md:31]: but this PR is not discussing the "meaning" of the attributes, only their encoding. Case-sensitivity is only a consideration when encoding. If you object to "Attribute keys are case-sensitive", wouldn't you also object to "the attribute key is a non-empty string"? But statements are about encoding.

**Review by AlexanderWert — APPROVED** [review]

File: specification/common/README.md (hunk 1 above)
Comment (MrAlias) [reply specification/common/README.md:31]: > Case-sensitivity is only a consideration when encoding.

Agreed. I think we should include this fact in the language being added.

[the remark this answers is not in the document]
Comment (MrAlias) [reply specification/common/README.md:31]: Sounds good to me, :+1:.

**Review by pyohannes — APPROVED** [review]

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
@MrAlias @reyang Any blocking feedback that still needs to be addressed?

**Comment by reyang** [comment]
> @MrAlias @reyang Any blocking feedback that still needs to be addressed?

Not from me 👍, thanks!

**Comment by carlosalberto** [comment]
Merging as there don't seem to be any blockers. Feel free to open an issue so we follow up any potential refinement @MrAlias, thanks!

**Review by MrAlias — APPROVED** [review]

**Pull request merged**